eda/data_analysis.py

```diff
@@ -104,9 +122,82 @@ def prepare_data(df: pd.DataFrame) -> pd.DataFrame:
+    Args:
+        df: Input DataFrame
+        category: Category for which spend data should be fetched
+        sector: Market sector for which spend data should be fetched
+
+    Returns:
+        Generate the correct date range to be used to train models.
+
+    """
+    category_sector_spend = df[(df["Category"] == category) & (df["MarketSector"] == sector)].reset_index(drop=True)
+    category_sector_spend.set_index("SpendMonth", inplace=True, drop=False)
+    # complete_data = pd.DataFrame(data=pd.date_range(start=category_sector_spend["SpendMonth"].min(axis=0),
+    #                                                 end=category_sector_spend["SpendMonth"].max(axis=0), freq="MS"),
+    #                              columns=["SpendMonth"])
+    end_date = max([category_sector_spend["SpendMonth"].max(axis=0), date.today() - relativedelta(months=2)])
+    complete_data = pd.DataFrame(data=pd.date_range(start=category_sector_spend["SpendMonth"].min(axis=0),
+                                                    end=end_date, freq="MS"), columns=["SpendMonth"])
+    complete_data["Category"] = category
+    complete_data["MarketSector"] = sector
+    complete_data["EvidencedSpend"] = 0.0
+    complete_data.set_index("SpendMonth", inplace=True, drop=False)
+    complete_data.update(category_sector_spend)
+    complete_data.reset_index(drop=True, inplace=True)
+    return complete_data
+
+
+def get_all_category_sector(latest: bool = False) -> list[Tuple[str, str]]:
+    category_sector_list = []
+    if not os.path.exists(local_category_sector_path):
+        latest = True
+    else:
+        df = pd.read_csv(local_category_sector_path)
+        category_sector_list = list(df.itertuples(index=False, name=None))
+
+    if latest:
+        sql = """
+            SELECT
+            COUNT (a.SpendMonth) AS RecordCount,a.Category,a.MarketSector
+            FROM
+            (
+               SELECT
+               DATEADD(month,3,CONVERT(date,CONCAT(spend.FYMonthKey,'01'),112)) as SpendMonth,
+               spend.Category,
+               ISNULL(cust.MarketSector,'Unassigned') AS MarketSector
+               FROM dbo.SpendAggregated AS spend
+               INNER JOIN dbo.FrameworkCategoryPillar frame ON spend.FrameworkNumber = frame.FrameworkNumber
+               LEFT JOIN dbo.Customers cust ON spend.CustomerURN = cust.CustomerKey
+               WHERE frame.STATUS IN ('Live', 'Expired - Data Still Received') AND DATEADD
+               (
+                  month,
+                  3,
+                  CONVERT(date,CONCAT(spend.FYMonthKey,'01'),112)
+               )
+               < DATEADD(month,-2,GETDATE())
+               GROUP BY spend.Category,
+               cust.MarketSector,
+               spend.FYMonthKey
+            )
+            AS a
+            GROUP BY a.Category,a.MarketSector
+            ORDER BY RecordCount DESC,a.Category,a.MarketSector
+            """
```

Comment:
Here we sort the combinations by record count. When I run this SQL query, the outcome seems to have combinations of MarketSector = Unassigned at the top. Is this expected? 

when I run this sql separately, the result looks like this:

"RecordCount","Category","MarketSector"
"205","Network Services","Unassigned"
"163","Financial Services","Unassigned"
"150","Document Management & Logistics","Unassigned"
"147","Technology Products & Services","Unassigned"
"130","Fleet","Unassigned"
"122","Workforce Health & Education","Unassigned"
"116","Construction","Defence and Security"
"116","Construction","Education"
"116","Construction","Government Policy"

Reply:
Good point, though it is not going to matter on the count of Category and MarketSector, but I modified the query.
Fixed!